### PR TITLE
BUG: Fix file-like objects failing in `read_xml` when iterparse is used

### LINKED ISSUE #50641: BUG:  pd.read_xml does not support file like object when iterparse is used
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
with open(file_, "r") as f:
    df = pd.read_xml(
        f,
        parser="etree",
        iterparse={root: list(elements)}
    )
```

### Issue Description

the method ```read_xml``` with iterparse as parms is used to read large xml file, but it's restricted to read only files on local disk.
this design is not justified since that the module ```iterparse``` from ```xml.etree.Elements``` does allow this features (reading from a file like object) 

```python
if (
    not isinstance(self.path_or_buffer, str) # -> condition that raise the Error
    or is_url(self.path_or_buffer)
    or is_fsspec_url(self.path_or_buffer)
    or self.path_or_buffer.startswith(("<?xml", "<"))
    or infer_compression(self.path_or_buffer, "infer") is not None
):
    raise ParserError(
        "iterparse is designed for large XML files that are fully extracted on "
        "local disk and not as compressed files or online sources."
    )
```
maybe something like this is better:
```python
if isinstance(self.path_or_buffer,str) or hasattr(self.path_or_buffer, "read"):
 #do something
else:
  raise ParserError(...)
```

### Expected Behavior

this must return a dataframe reither than raise Exception

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 87cfe4e38bafe7300a6003a1d18bd80f3f77c763
python           : 3.10.4.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-136-generic
Version          : #153-Ubuntu SMP Thu Nov 24 15:56:58 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.5.0
numpy            : 1.23.4
pytz             : 2022.5
dateutil         : 2.8.2
setuptools       : 58.1.0
pip              : 22.0.4
Cython           : None
pytest           : 7.1.3
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : 
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
tzdata           : None

</details>

### PR DESCRIPTION
**Pull request by ParfaitG**
- [X] closes #50641
- [X] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [X] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [X] Added an entry in the latest `doc/source/whatsnew/v2.0.0.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/whatsnew/v1.5.3.rst [hunk 1]
```diff
@@ -33,6 +33,7 @@ Bug fixes
 - Bug when chaining several :meth:`.Styler.concat` calls, only the last styler was concatenated (:issue:`49207`)
 - Fixed bug when instantiating a :class:`DataFrame` subclass inheriting from ``typing.Generic`` that triggered a ``UserWarning`` on python 3.11 (:issue:`49649`)
 - Bug in :func:`pandas.testing.assert_series_equal` (and equivalent ``assert_`` functions) when having nested data and using numpy >= 1.25 (:issue:`50360`)
+- Bug in :func:`read_xml` where file-like objects failed when iterparse is used (:issue:`50641`)
```
Comment (phofl) [inline doc/source/whatsnew/v1.5.3.rst:36]: Can we put this under I/O?

Comment (ParfaitG) [reply doc/source/whatsnew/v1.5.3.rst:36]: I can add this note to 2.0. I mixed up the versions.

File: pandas/tests/io/xml/test_xml.py [hunk 2]
```diff
@@ -1351,18 +1351,66 @@ def test_string_error(parser):
...
+            return None
+        else:
+            df_filelike = read_xml(
                 f,
                 parser=parser,
                 iterparse={"book": ["category", "title", "year", "author", "price"]},
             )
 
+    IOfunc = StringIO if mode == "r" else BytesIO
+    with open(filename, mode) as f:
```
Comment (mroeschke) [inline pandas/tests/io/xml/test_xml.py:1378]: Could you make this new block a separate test?

Comment (ParfaitG) [reply pandas/tests/io/xml/test_xml.py:1394]: Sounds good. Will do.

**Review by mroeschke — APPROVED** [review]

**Pull request merged**